### ISSUE TITLE
Performance issue with SQLite offline store

### ISSUE THREAD

**Issue by Jan-NET** [Azure/azure-mobile-apps#575]
We run a Xamarin App with Microsoft.Azure.Mobile.Client for years now. Thank you for this great sync framework!
The last weeks we migrated this app with Microsoft.Azure.Mobile.Client 4.1.2 in place to MAUI and Microsoft.Datasync.Client 5.0.20. 
Now we have a tremendous performance problem.  The migrated app syncs the data (about 100000 data records distributed over 17 tables) about 20 times slower than the old app (iOS and Android)!

So we went a step back and experimented with your Todo sample app (https://github.com/Azure/azure-mobile-apps/tree/main/samples/TodoApp)/TodoApp.MAUI), upgraded to the offline store and take some measurements for the sync process (always 10000 inserts into TodoItem table, same emulator): If we sum up the Android millis (X-Android-Received-Millis minus X-Android-Sent-Millis) from the response headers we get plausible response time values. So we focused on the client side.
If we execute 10000 times await _service.SaveItemAsync(item) (no sync to the API!) we measured about 5 minutes execution time! Not good! Even the SQlite-Flag cache=shared does not lead to any improvements.
In the next experimental step we followed the guideline https://learn.microsoft.com/en-us/dotnet/maui/data-cloud/database-sqlite?view=net-maui-7.0, used the sqlite-net-pcl NuGet package and upgraded the TodoItem sample app with a second local db (new SQLiteAsyncConnection) and executed again 10000 inserts. This is done in about 1 minute. Ah, much better!
If we EnableWriteAheadLoggingAsync() we can improve to 17 seconds. Great!
So, 17 seconds vs. 300 (5 minutes)! If we are very, very generous that’s the “about 20 times slower”.

So, what is going on with the Microsoft.Datasync.Client? What are the tricks? Or can´t we see the forest for the trees?

## To Reproduce

1. Use https://github.com/Azure/azure-mobile-apps/tree/main/samples/TodoApp)/TodoApp.MAUI
2. Activate the offline store
3. Insert 10.000 TodoItem (no sync to API needed)
4. Measure the time

## Expected behavior

Insert 10.000 TodoItems in < 60 seconds.

## What platforms?

* Server: not relevant

* Client: 
- Microsoft.Datasync.Client 5.0.20 
- .NET 6
-  sqlite-net-pcl 1.8.116
- SQLitePCLRaw.bundle_green 2.1.3

**Comment by adrianhall** [comment]
You are missing the requirement around transactions and rollbacks when we commit a change across multiple threads.  That being said, there is definite improvements we can do.  I'm happy to take a PR for the SQLiteOfflineStore if you bump into something that can be optimized prior to me getting to it.

**Comment by amertlich** [comment]
The `OfflineSqliteStore` has an `UpsertAsync` method that we've leveraged to handle local batch transactions.  The only issue is that this doesn't track the individual operations in the __operation queue table.  So we've had to manually batch insert these transactions as well using the `ExecuteQueryAsync` method that also exists we can wrap this all in a `BEGIN TRANSACTION` and manually execute the necessary SQL on the client database.

Now the only problem is that in memory state of the [OperationQueue ](https://github.com/Azure/azure-mobile-apps/blob/main/sdk/dotnet/src/Microsoft.Datasync.Client/Offline/Queue/OperationsQueue.cs)(which is an internal class) has the wrong `SequenceId` since we just manually inserted into the queue.  The quick and dirty way I've found to reinitialize this is by closing the storage and re-opening it which is still faster than inserting each item manually.

So basically, you have to write your own local batch insertion/update logic -- it's technically possible, but requires a bit of a work around.  It would be nice if there were batch operation overloads just for the local store only, even if they're still synchronized individually.

**Comment by Jan-NET** [comment]
But the transactions/rollbacks-concept was used in Microsoft.Azure.Mobile.Client 4.*.* too?
I do not understand the huge performance differences old vs. new, Microsoft.Azure.Mobile.Client vs. Microsoft.Datasync.Client! 

I thought I could combine the new and old sync client into one sample app for an at-a-glance performance comparison. But unfortunately I can't compile that, so I'll try to provide two separate sample apps tomorrow.

**Comment by adrianhall** [comment]
transactions / rollback was NOT used in the old client.

The SQLiteStore for the new client is basically a copy of the old store client with an updated SQLite engine.  It should not be drastically different performance wise.  I need to dig into profiling data to understand why this is the case.

**Comment by Jan-NET** [comment]
Oh, wrong assumption in my part. So, this inner breaking change could have an impact.

The weird thing: As I said, I created today two lean XamarinForms apps, one with the old client, one with the new client. The comparison shows no significant performance differences, the new client is about 10% slower than the old one!
(So it makes no sense to provide the XF sample apps to GIT.)

I have to go back and puzzle out our whole sync process.
However, the realization, that the package sqlite-net-pcl 1.8.116 can insert data more quickly, remains. I guess, the transactions/rollbacks-concept will not double or triple the insert time.

**Comment by Jan-NET** [comment]
Today I go further with my test apps and synced with a backend (20000 records in TodoItem table in Azure SQL).
The backend for both test apps is the TodoAppService.NET6 from your Todoitem sample (pagesize = 50). 

Sync time Microsoft.Azure.Mobile.Client: about 2 minutes.
Sync time Microsoft.Datasync.Client: about 12 minutes.

With the findings from yesterday in mind, that the local inserts for the new client are only 10% slower than the old client: Where is the time lost? HTTP-Response processing? Serialization? 
The first GET-requests for both clients from the protocol stream in Azure portal:

Old client:
```
2023-02-03 14:06:42.536 +00:00 GET https://*.net/tables/TodoItem?$filter=(updatedAt%20ge%20datetimeoffset'1970-01-01T00%3A00%3A00.0000000%2B00%3A00')&$orderby=updatedAt&$skip=0&$top=50&__includeDeleted=true 
2023-02-03 14:06:44.290 +00:00 GET https://*.net/tables/TodoItem?$filter=(updatedAt%20ge%20datetimeoffset'1970-01-01T00%3A00%3A00.0000000%2B00%3A00')&$orderby=updatedAt&$skip=0&$top=50&__includeDeleted=true 
2023-02-03 14:06:44.442 +00:00 GET https://*.net/tables/TodoItem?$filter=(updatedAt%20ge%20datetimeoffset'2023-01-30T09%3A31%3A00.7050000%2B00%3A00')&$orderby=updatedAt&$skip=0&$top=50&__includeDeleted=true 
2023-02-03 14:06:44.520 +00:00 GET https://*.net/tables/TodoItem?$filter=(updatedAt%20ge%20datetimeoffset'2023-01-30T09%3A31%3A00.7050000%2B00%3A00')&$orderby=updatedAt&$skip=0&$top=50&__includeDeleted=true 
2023-02-03 14:06:44.626 +00:00 GET https://*.net/tables/TodoItem?$filter=(updatedAt%20ge%20datetimeoffset'2023-01-30T09%3A31%3A01.0330000%2B00%3A00')&$orderby=updatedAt&$skip=0&$top=50&__includeDeleted=true 
2023-02-03 14:06:44.695 +00:00 GET https://*.net/tables/TodoItem?$filter=(updatedAt%20ge%20datetimeoffset'2023-01-30T09%3A31%3A01.0330000%2B00%3A00')&$orderby=updatedAt&$skip=0&$top=50&__includeDeleted=true
``` 
New client:
```
2023-02-03 12:22:19.911 +00:00 GET https://*.net/tables/todoitem?$filter=(updatedAt%20gt%20cast(1970-01-01T00:00:00.000Z,Edm.DateTimeOffset))&$orderby=updatedAt&$count=true&__includedeleted=true 
2023-02-03 12:22:23.958 +00:00 GET https://*.net/tables/todoitem?$filter=(updatedAt%20gt%20cast(1970-01-01T00:00:00.000Z,Edm.DateTimeOffset))&$orderby=updatedAt&$count=true&__includedeleted=true 
2023-02-03 12:22:26.200 +00:00 GET https://*.net/tables/todoitem?$filter=(updatedAt%20gt%20cast(1970-01-01T00:00:00.000Z,Edm.DateTimeOffset))&$orderby=updatedAt&$count=true&__includedeleted=true&$skip=50 
2023-02-03 12:22:26.490 +00:00 GET https://*.net/tables/todoitem?$filter=(updatedAt%20gt%20cast(1970-01-01T00:00:00.000Z,Edm.DateTimeOffset))&$orderby=updatedAt&$count=true&__includedeleted=true&$skip=50 
2023-02-03 12:22:28.133 +00:00 GET https://*.net/tables/todoitem?$filter=(updatedAt%20gt%20cast(1970-01-01T00:00:00.000Z,Edm.DateTimeOffset))&$orderby=updatedAt&$count=true&__includedeleted=true&$skip=100 
2023-02-03 12:22:28.568 +00:00 GET https://*.net/tables/todoitem?$filter=(updatedAt%20gt%20cast(1970-01-01T00:00:00.000Z,Edm.DateTimeOffset))&$orderby=updatedAt&$count=true&__includedeleted=true&$skip=100 
2023-02-03 12:22:30.144 +00:00 GET https://*.net/tables/todoitem?$filter=(updatedAt%20gt%20cast(1970-01-01T00:00:00.000Z,Edm.DateTimeOffset))&$orderby=updatedAt&$count=true&__includedeleted=true&$skip=150 
```


I don´t know why there is the GET-redundancy for both clients, but you can see the long timespans between the requests for the new client in comparison to the short timespans for the old client.

Any suggestions how I can dig in further?

**Comment by Jan-NET** [comment]
If you mentally split up 
`public async Task PullItemsAsync(string tableName, string query, PullOptions options, CancellationToken cancellationToken = default)` in [https://github.com/Azure/azure-mobile-apps/blob/main/sdk/dotnet/src/Microsoft.Datasync.Client/Offline/SyncContext.cs](url)
in the four phases

1. HTTP-request phase (code line 334)
2. "pre store item"-phase (code lines 336 to 360)
3. "store item"-phase (code lines 361 to 371)
4. "post store item"-phase (code lines 372 to 376, + line 360)

my measurements show an average time weighting between the four phases from 7:19:52:22 (P1:P2:P3:P4, sum up to 100%). 
These values vary from run to run and device to device but are a first orientation.

22% of the execution time for the “post store item"-phase caught my attention, because it is only 
```
if (updatedAt.HasValue)
{
   await DeltaTokenStore.SetDeltaTokenAsync(tableName, queryId, updatedAt.Value, cancellationToken).ConfigureAwait(false);
}
```
and the associated line 360 from above:

`   DateTimeOffset? updatedAt = ServiceSerializer.GetUpdatedAt(item)?.ToUniversalTime();`


Why is `DeltaTokenStore.SetDeltaTokenAsync `executed for every item in the loop? If the loop is successful, one execution is sufficient, isn’t it? 
In my understanding the refreshed DeltaToken will not be used in the next GET of the foreach-loop, due to the use of SKIP in the GET-requests. If the loop is erroneous, yes, then we have the drawback to pull all items again. A parallel pull for the same table (is it even possible?) has the same drawback.

In the "pre store item"-phase the whole execution time is consumed by line 355:
`var pendingOperation = await OperationsQueue.GetOperationByItemIdAsync(tableName, itemId, cancellationToken).ConfigureAwait(false);`
Make it sense to load the pending operations for the table up front (before the foreach loop) in an e.g. hashtable?

In the "store item"-phase `await EnsureInitializedAsync(cancellationToken).ConfigureAwait(false);`, line 222 in [https://github.com/Azure/azure-mobile-apps/blob/main/sdk/dotnet/src/Microsoft.Datasync.Client.SQLiteStore/OfflineSQLiteStore.cs](url), causes high costs. The `UpsertAsync `method is in my opinion more focused on one shot. For a loop of upserts it would be sufficient to call EnsureInitializedAsync once?!

So, there is not the one and only joker, but many individual measures.

**Comment by adrianhall** [comment]
To answer your specific questions:

1) SetDeltaTokenAsync is called on each iteration to ensure that if the pull operation is interrupted mid-sequence, the delta-token is maintained.  This ensures there are no conflicts after the fact.  Since we use an IAsyncEnumerable with multiple GET calls behind it, we don't necessarily know when the problem occurs.  This is primarily a mobile concern, but there are significant consistency problems in real-world environments when we don't store the delta-token immediately.

2) Loading the operation queue into memory, while it seems like a good idea, is NOT a good idea when the operation queue gets large.  Having said that, there are definite optimizations we can do there, such as a memory restricted but SQL backed operation queue.  This is why the operations queue is completely abstracted.

3) We can certainly improve EnsureInitializedAsync() to check IsInitialized first and do an early return (thereby losing the async lock requirement in the majority of cases). 

Since you have been playing with timing, why don't you do #3 here, and see what effect it has on timing (my suspicion is that it will be minimal).  If it's effective, do a PR as I view this as a non-breaking change.

On another note, I'm intending on starting with a new sync engine design.  If you are interested in taking part, let me know.

**Comment by adrianhall** [comment]
Another thought - maybe we can do something whereby we set a "number of iterations before we set the delta-token" - in highly known environments (e.g. desktops), we can write the delta-token every 25 saves, resulting in a significant reduction in the update time.   That should be relatively easy to set up with an additional optional setting and non-breaking change.

**Comment by adrianhall** [comment]
I've just merged a change that prevents an async lock from being held if the store is already initialized - this will directly hit the 22% phase 1 that you mentioned earlier.  Can you check that the performance is actually impacted as expected there?

Next up, I think we should write a new integration test that does a 25K insert via synchronization - that will allow us to see if we are really impacting performance in the store.

**Comment by adrianhall** [comment]
Let's break this down and set up some issues for each proposed solution - we can then go about building something.

For the new sync engine - nothing has been started yet.  I'd like to get an initial design document done, and then I can post that and get commentary on it.  There will be a new repository (at least temporarily) while we go over initial design.

**Comment by adrianhall** [comment]
Given all the current items are done, and we have non-5.1.0 issues filed, plus a re-vamp of the sync engine going on, I'm going to call this complete at this point.

**Comment by Jan-NET** [comment]
Additional comment: After my poor test for #575 in an emulator I decided to make some runs with a standalone android tablet.

The sync time median over 10+ runs for each client version (always pull 20K from same backend) is quite interesting. 
Version `5.0.21` is about 3 % slower than `5.0.20` (These are variances in transfer rates or backend processing, not a thing.). 
But version `5.1.0` is 60% faster than 5.0.21! 
Wow, that is astonishing! If I change `WriteDeltaTokenInterval` from `default 1` to `25` an additional improvement of 13% is detectable.

What are the key points for this strong improvement?

**Comment by adrianhall** [comment]
Profiling the code, the EnsureInitializedAsync lock took the majority of the time, especially in the multi-threaded version.  Adding the inner and outer checks to make sure that InitializeAsync() was called precisely once was the big win.

The next biggest thing is the double transaction write for operations queue and item, but that requires a re-architecture of the store (which is now underway).

**Comment by danbridgellc** [comment]
@Jan-NET - I know this is off topic a bit but this is best thread I have found so far...in your tests, have you seen an issue while initializing the SQLite offline store where it does not create the table?  I have been using 5.0.21 since it came out and it has been working fine.  After upgrading to 5.1.0, it still worked fine...until I deleted my local SQLite database and went to sync the first time.  It throws SQLite exception  saying it cannot prepare sql statement.  I open the database and sure enough the table is not there.  I simply downgrade packages to 5.0.21 and retried and it worked fine.  Wondering if I missed a change to the library or if I log a new bug with repo steps?

**Comment by adrianhall** [comment]
The tests actually test that creation flow, so it's explicitly passing in 5.1.0 and 5.1.1.  

First step - add an ILogger to the OfflineSQLiteStore to log the SQL statements - make sure the SQL is correct for your table creation.

Second step - open an issue with as much detail as possible so we can try to reproduce it.

**Comment by Jan-NET** [comment]
I've never stumbled across such a problem and had many delete/create scenarios over the past few weeks (using 5.0.20/5.0.21/5.1.0). So, take Adrian's advice, that would be a good start.
